File: gallery/pdf_writer/src/jpeg/bin/gallery/pdf_writer/src/jpeg/DCT.go
package main
import (
)

type GoNullable struct {
  value interface{}
  has_value bool
}

type IDCT struct { 
  cosTable []int64 `json:"cosTable"` 
  zigzagMap []int64 `json:"zigzagMap"` 
}

func CreateNew_IDCT() *IDCT {
  me := new(IDCT)
  me.cosTable = 
  make([]int64, int64(64))
  
  me.zigzagMap = 
  make([]int64, int64(64))
  
  me.cosTable[int64(0)] = int64(1024)
  me.cosTable[int64(1)] = int64(1004)
  me.cosTable[int64(2)] = int64(946)
  me.cosTable[int64(3)] = int64(851)
  me.cosTable[int64(4)] = int64(724)
  me.cosTable[int64(5)] = int64(569)
  me.cosTable[int64(6)] = int64(392)
  me.cosTable[int64(7)] = int64(200)
  me.cosTable[int64(8)] = int64(1024)
  me.cosTable[int64(9)] = int64(851)
  me.cosTable[int64(10)] = int64(392)
  me.cosTable[int64(11)] = int64(-200)
  me.cosTable[int64(12)] = int64(-724)
  me.cosTable[int64(13)] = int64(-1004)
  me.cosTable[int64(14)] = int64(-946)
  me.cosTable[int64(15)] = int64(-569)
  me.cosTable[int64(16)] = int64(1024)
  me.cosTable[int64(17)] = int64(569)
  me.cosTable[int64(18)] = int64(-392)
  me.cosTable[int64(19)] = int64(-1004)
  me.cosTable[int64(20)] = int64(-724)
  me.cosTable[int64(21)] = int64(200)
  me.cosTable[int64(22)] = int64(946)
  me.cosTable[int64(23)] = int64(851)
  me.cosTable[int64(24)] = int64(1024)
  me.cosTable[int64(25)] = int64(200)
  me.cosTable[int64(26)] = int64(-946)
  me.cosTable[int64(27)] = int64(-569)
  me.cosTable[int64(28)] = int64(724)
  me.cosTable[int64(29)] = int64(851)
  me.cosTable[int64(30)] = int64(-392)
  me.cosTable[int64(31)] = int64(-1004)
  me.cosTable[int64(32)] = int64(1024)
  me.cosTable[int64(33)] = int64(-200)
  me.cosTable[int64(34)] = int64(-946)
  me.cosTable[int64(35)] = int64(569)
  me.cosTable[int64(36)] = int64(724)
  me.cosTable[int64(37)] = int64(-851)
  me.cosTable[int64(38)] = int64(-392)
  me.cosTable[int64(39)] = int64(1004)
  me.cosTable[int64(40)] = int64(1024)
  me.cosTable[int64(41)] = int64(-569)
  me.cosTable[int64(42)] = int64(-392)
  me.cosTable[int64(43)] = int64(1004)
  me.cosTable[int64(44)] = int64(-724)
  me.cosTable[int64(45)] = int64(-200)
  me.cosTable[int64(46)] = int64(946)
  me.cosTable[int64(47)] = int64(-851)
  me.cosTable[int64(48)] = int64(1024)
  me.cosTable[int64(49)] = int64(-851)
  me.cosTable[int64(50)] = int64(392)
  me.cosTable[int64(51)] = int64(200)
  me.cosTable[int64(52)] = int64(-724)
  me.cosTable[int64(53)] = int64(1004)
  me.cosTable[int64(54)] = int64(-946)
  me.cosTable[int64(55)] = int64(569)
  me.cosTable[int64(56)] = int64(1024)
  me.cosTable[int64(57)] = int64(-1004)
  me.cosTable[int64(58)] = int64(946)
  me.cosTable[int64(59)] = int64(-851)
  me.cosTable[int64(60)] = int64(724)
  me.cosTable[int64(61)] = int64(-569)
  me.cosTable[int64(62)] = int64(392)
  me.cosTable[int64(63)] = int64(-200)
  me.zigzagMap[int64(0)] = int64(0)
  me.zigzagMap[int64(1)] = int64(1)
  me.zigzagMap[int64(2)] = int64(8)
  me.zigzagMap[int64(3)] = int64(16)
  me.zigzagMap[int64(4)] = int64(9)
  me.zigzagMap[int64(5)] = int64(2)
  me.zigzagMap[int64(6)] = int64(3)
  me.zigzagMap[int64(7)] = int64(10)
  me.zigzagMap[int64(8)] = int64(17)
  me.zigzagMap[int64(9)] = int64(24)
  me.zigzagMap[int64(10)] = int64(32)
  me.zigzagMap[int64(11)] = int64(25)
  me.zigzagMap[int64(12)] = int64(18)
  me.zigzagMap[int64(13)] = int64(11)
  me.zigzagMap[int64(14)] = int64(4)
  me.zigzagMap[int64(15)] = int64(5)
  me.zigzagMap[int64(16)] = int64(12)
  me.zigzagMap[int64(17)] = int64(19)
  me.zigzagMap[int64(18)] = int64(26)
  me.zigzagMap[int64(19)] = int64(33)
  me.zigzagMap[int64(20)] = int64(40)
  me.zigzagMap[int64(21)] = int64(48)
  me.zigzagMap[int64(22)] = int64(41)
  me.zigzagMap[int64(23)] = int64(34)
  me.zigzagMap[int64(24)] = int64(27)
  me.zigzagMap[int64(25)] = int64(20)
  me.zigzagMap[int64(26)] = int64(13)
  me.zigzagMap[int64(27)] = int64(6)
  me.zigzagMap[int64(28)] = int64(7)
  me.zigzagMap[int64(29)] = int64(14)
  me.zigzagMap[int64(30)] = int64(21)
  me.zigzagMap[int64(31)] = int64(28)
  me.zigzagMap[int64(32)] = int64(35)
  me.zigzagMap[int64(33)] = int64(42)
  me.zigzagMap[int64(34)] = int64(49)
  me.zigzagMap[int64(35)] = int64(56)
  me.zigzagMap[int64(36)] = int64(57)
  me.zigzagMap[int64(37)] = int64(50)
  me.zigzagMap[int64(38)] = int64(43)
  me.zigzagMap[int64(39)] = int64(36)
  me.zigzagMap[int64(40)] = int64(29)
  me.zigzagMap[int64(41)] = int64(22)
  me.zigzagMap[int64(42)] = int64(15)
  me.zigzagMap[int64(43)] = int64(23)
  me.zigzagMap[int64(44)] = int64(30)
  me.zigzagMap[int64(45)] = int64(37)
  me.zigzagMap[int64(46)] = int64(44)
  me.zigzagMap[int64(47)] = int64(51)
  me.zigzagMap[int64(48)] = int64(58)
  me.zigzagMap[int64(49)] = int64(59)
  me.zigzagMap[int64(50)] = int64(52)
  me.zigzagMap[int64(51)] = int64(45)
  me.zigzagMap[int64(52)] = int64(38)
  me.zigzagMap[int64(53)] = int64(31)
  me.zigzagMap[int64(54)] = int64(39)
  me.zigzagMap[int64(55)] = int64(46)
  me.zigzagMap[int64(56)] = int64(53)
  me.zigzagMap[int64(57)] = int64(60)
  me.zigzagMap[int64(58)] = int64(61)
  me.zigzagMap[int64(59)] = int64(54)
  me.zigzagMap[int64(60)] = int64(47)
  me.zigzagMap[int64(61)] = int64(55)
  me.zigzagMap[int64(62)] = int64(62)
  me.zigzagMap[int64(63)] = int64(63)
  return me;
}
func (this *IDCT) dezigzag (zigzag []int64) []int64 {
  var block []int64= make([]int64, int64(64));
  var i int64= int64(0);
  for i < int64(64) {
    var pos int64= this.zigzagMap[i];
    var val int64= zigzag[i];
    block[pos] = val
    i = i + int64(1); 
  }
  return block
}
func (this *IDCT) idct1d (input []int64, startIdx int64, stride int64, output []int64, outIdx int64, outStride int64) () {
  var x int64= int64(0);
  for x < int64(8) {
    var sum int64= int64(0);
    var u int64= int64(0);
    for u < int64(8) {
      var coeff int64= input[(startIdx + (u * stride))];
      if  coeff != int64(0) {
        var cosVal int64= this.cosTable[((x * int64(8)) + u)];
        var contrib int64= coeff * cosVal;
        if  u == int64(0) {
          contrib = int64((contrib * int64(724)) >> uint(int64(10))); 
        }
        sum = sum + contrib; 
      }
      u = u + int64(1); 
    }
    output[outIdx + (x * outStride)] = int64(sum >> uint(int64(11)))
    x = x + int64(1); 
  }
}
func (this *IDCT) transform (block []int64, output []int64) () {
  var temp []int64= make([]int64, int64(64));
  var row int64= int64(0);
  for row < int64(8) {
    var rowStart int64= row * int64(8);
    this.idct1d(block, rowStart, int64(1), temp, rowStart, int64(1));
    row = row + int64(1); 
  }
  var col int64= int64(0);
  for col < int64(8) {
    this.idct1d(temp, col, int64(8), output, col, int64(8));
    col = col + int64(1); 
  }
  var i int64= int64(0);
  for i < int64(64) {
    var val int64= (output[i]) + int64(128);
    if  val < int64(0) {
      val = int64(0); 
    }
    if  val > int64(255) {
      val = int64(255); 
    }
    output[i] = val
    i = i + int64(1); 
  }
}
func (this *IDCT) transformFast (coeffs []int64, output []int64) () {
  this.transform(coeffs, output);
}

File: gallery/pdf_writer/src/jpeg/bin/gallery/pdf_writer/src/jpeg/DCT.js
class IDCT  {
  constructor() {
    this.cosTable = new Int32Array(64);
    this.zigzagMap = new Int32Array(64);
    this.cosTable[0] = 1024;
    this.cosTable[1] = 1004;
    this.cosTable[2] = 946;
    this.cosTable[3] = 851;
    this.cosTable[4] = 724;
    this.cosTable[5] = 569;
    this.cosTable[6] = 392;
    this.cosTable[7] = 200;
    this.cosTable[8] = 1024;
    this.cosTable[9] = 851;
    this.cosTable[10] = 392;
    this.cosTable[11] = -200;
    this.cosTable[12] = -724;
    this.cosTable[13] = -1004;
    this.cosTable[14] = -946;
    this.cosTable[15] = -569;
    this.cosTable[16] = 1024;
    this.cosTable[17] = 569;
    this.cosTable[18] = -392;
    this.cosTable[19] = -1004;
    this.cosTable[20] = -724;
    this.cosTable[21] = 200;
    this.cosTable[22] = 946;
    this.cosTable[23] = 851;
    this.cosTable[24] = 1024;
    this.cosTable[25] = 200;
    this.cosTable[26] = -946;
    this.cosTable[27] = -569;
    this.cosTable[28] = 724;
    this.cosTable[29] = 851;
    this.cosTable[30] = -392;
    this.cosTable[31] = -1004;
    this.cosTable[32] = 1024;
    this.cosTable[33] = -200;
    this.cosTable[34] = -946;
    this.cosTable[35] = 569;
    this.cosTable[36] = 724;
    this.cosTable[37] = -851;
    this.cosTable[38] = -392;
    this.cosTable[39] = 1004;
    this.cosTable[40] = 1024;
    this.cosTable[41] = -569;
    this.cosTable[42] = -392;
    this.cosTable[43] = 1004;
    this.cosTable[44] = -724;
    this.cosTable[45] = -200;
    this.cosTable[46] = 946;
    this.cosTable[47] = -851;
    this.cosTable[48] = 1024;
    this.cosTable[49] = -851;
    this.cosTable[50] = 392;
    this.cosTable[51] = 200;
    this.cosTable[52] = -724;
    this.cosTable[53] = 1004;
    this.cosTable[54] = -946;
    this.cosTable[55] = 569;
    this.cosTable[56] = 1024;
    this.cosTable[57] = -1004;
    this.cosTable[58] = 946;
    this.cosTable[59] = -851;
    this.cosTable[60] = 724;
    this.cosTable[61] = -569;
    this.cosTable[62] = 392;
    this.cosTable[63] = -200;
    this.zigzagMap[0] = 0;
    this.zigzagMap[1] = 1;
    this.zigzagMap[2] = 8;
    this.zigzagMap[3] = 16;
    this.zigzagMap[4] = 9;
    this.zigzagMap[5] = 2;
    this.zigzagMap[6] = 3;
    this.zigzagMap[7] = 10;
    this.zigzagMap[8] = 17;
    this.zigzagMap[9] = 24;
    this.zigzagMap[10] = 32;
    this.zigzagMap[11] = 25;
    this.zigzagMap[12] = 18;
    this.zigzagMap[13] = 11;
    this.zigzagMap[14] = 4;
    this.zigzagMap[15] = 5;
    this.zigzagMap[16] = 12;
    this.zigzagMap[17] = 19;
    this.zigzagMap[18] = 26;
    this.zigzagMap[19] = 33;
    this.zigzagMap[20] = 40;
    this.zigzagMap[21] = 48;
    this.zigzagMap[22] = 41;
    this.zigzagMap[23] = 34;
    this.zigzagMap[24] = 27;
    this.zigzagMap[25] = 20;
    this.zigzagMap[26] = 13;
    this.zigzagMap[27] = 6;
    this.zigzagMap[28] = 7;
    this.zigzagMap[29] = 14;
    this.zigzagMap[30] = 21;
    this.zigzagMap[31] = 28;
    this.zigzagMap[32] = 35;
    this.zigzagMap[33] = 42;
    this.zigzagMap[34] = 49;
    this.zigzagMap[35] = 56;
    this.zigzagMap[36] = 57;
    this.zigzagMap[37] = 50;
    this.zigzagMap[38] = 43;
    this.zigzagMap[39] = 36;
    this.zigzagMap[40] = 29;
    this.zigzagMap[41] = 22;
    this.zigzagMap[42] = 15;
    this.zigzagMap[43] = 23;
    this.zigzagMap[44] = 30;
    this.zigzagMap[45] = 37;
    this.zigzagMap[46] = 44;
    this.zigzagMap[47] = 51;
    this.zigzagMap[48] = 58;
    this.zigzagMap[49] = 59;
    this.zigzagMap[50] = 52;
    this.zigzagMap[51] = 45;
    this.zigzagMap[52] = 38;
    this.zigzagMap[53] = 31;
    this.zigzagMap[54] = 39;
    this.zigzagMap[55] = 46;
    this.zigzagMap[56] = 53;
    this.zigzagMap[57] = 60;
    this.zigzagMap[58] = 61;
    this.zigzagMap[59] = 54;
    this.zigzagMap[60] = 47;
    this.zigzagMap[61] = 55;
    this.zigzagMap[62] = 62;
    this.zigzagMap[63] = 63;
  }
  dezigzag (zigzag) {
    const block = new Int32Array(64);
    let i = 0;
    while (i < 64) {
      const pos = this.zigzagMap[i];
      const val = zigzag[i];
      block[pos] = val;
      i = i + 1;
    };
    return block;
  };
  idct1d (input, startIdx, stride, output, outIdx, outStride) {
    let x = 0;
    while (x < 8) {
      let sum = 0;
      let u = 0;
      while (u < 8) {
        const coeff = input[(startIdx + (u * stride))];
        if ( coeff != 0 ) {
          const cosVal = this.cosTable[((x * 8) + u)];
          let contrib = coeff * cosVal;
          if ( u == 0 ) {
            contrib = ((contrib * 724) >> 10);
          }
          sum = sum + contrib;
        }
        u = u + 1;
      };
      output[outIdx + (x * outStride)] = (sum >> 11);
      x = x + 1;
    };
  };
  transform (block, output) {
    const temp = new Int32Array(64);
    let row = 0;
    while (row < 8) {
      const rowStart = row * 8;
      this.idct1d(block, rowStart, 1, temp, rowStart, 1);
      row = row + 1;
    };
    let col = 0;
    while (col < 8) {
      this.idct1d(temp, col, 8, output, col, 8);
      col = col + 1;
    };
    let i = 0;
    while (i < 64) {
      let val = (output[i]) + 128;
      if ( val < 0 ) {
        val = 0;
      }
      if ( val > 255 ) {
        val = 255;
      }
      output[i] = val;
      i = i + 1;
    };
  };
  transformFast (coeffs, output) {
    this.transform(coeffs, output);
  };
}
